### PR TITLE
osd: propogate mlcod to replicas and fix problems with read from replica

### PR DESCRIPTION
**Pull request by athanatos**

### CODE REVIEW

**Comment by athanatos** [comment]
http://pulpito.ceph.com/sjust-2019-12-20_06:42:20-rados-wip-sjust-testing-2019-12-19-distro-basic-smithi/

Failures appear unrelated/similar to other master runs.

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -11266,14 +11266,19 @@ int PrimaryLogPG::find_object_context(const hobject_t& oid,
...
+	dout(20) << __func__ << " clone is recovering " << soid << dendl;
+	return -EAGAIN;
+      } else {
+	dout(20) << __func__ << " missing clone " << soid << dendl;
+	return -ENOENT;
+      }
     } else {
-      dout(20) << __func__ << " missing clone " << soid << dendl;
+      dout(20) << __func__ << " replica missing clone" << soid << dendl;
       return -ENOENT;
```
Comment (liewegas) [inline src/osd/PrimaryLogPG.cc]: should the caller (do_op?) then also bounce the op to the primary if it gets ENOENT?

Comment (athanatos) [reply src/osd/PrimaryLogPG.cc]: Yeah, it does.  The problem was that the is_degrade_or_backfilling_object check asserts that some primary only things are populated.

Comment (athanatos) [reply src/osd/PrimaryLogPG.cc]: See PrimaryLogPG::maybe_handle_cache_detail

**Pull request merged**